File: tree/trie/208_implement_trie.cc
#include <iostream>
#include <string>
#include <vector>

// 2023/12/22: 直接拷贝过来的, 从评论来看, 考的频率较高, 需要注意.
class Trie {
 public:
  Trie() : children_(26), is_end_(false) {}

  void Insert(const std::string& word) {
    Trie* node = this;
    for (char ch : word) {
      ch -= 'a';
      if (node->children_[ch] == nullptr) {
        node->children_[ch] = new Trie();
      }
      node = node->children_[ch];
    }
    node->is_end_ = true;
  }

  bool Search(const std::string& word) {
    Trie* node = this->SearchPrefix(word);
    return node != nullptr && node->is_end_;
  }

  bool StartWith(const std::string& prefix) {
    return this->SearchPrefix(prefix) != nullptr;
  }

 private:
  Trie* SearchPrefix(const std::string& prefix) {
    Trie* node = this;
    for (char ch : prefix) {
      ch -= 'a';
      if (node->children_[ch] == nullptr) {
        return nullptr;
      }
      node = node->children_[ch];
    }
    return node;
  }

  std::vector<Trie*> children_;
  bool is_end_;
};

int main(int argc, char* argv[]) {
  Trie trie;
  trie.Insert("apple");
  std::cout << "search apple: " << trie.Search("apple") << std::endl;
  std::cout << "search app: " << trie.Search("app") << std::endl;
  std::cout << "start with app: " << trie.StartWith("app") << std::endl;
  trie.Insert("app");
  std::cout << "search app: " << trie.Search("app") << std::endl;
  return 0;
}

File: tree/trie/208_implement_trie.py
# 2026/05/21: 直接抄答案
class Trie:

    def __init__(self):
        self.children = [None] * 26  # 注: 用 None 进行初始化.
        self.is_end = False

    def insert(self, word: str) -> None:
        node = self  # 用 self 来初始化 node.
        for ch in word:
            ch = ord(ch) - ord("a")
            if not node.children[ch]:
                node.children[ch] = Trie()
            node = node.children[ch]
        node.is_end = True

    def search_prefix(self, prefix: str) -> "Trie":  # 这里的返回类型需要前向引用
        node = self
        for ch in prefix:
            ch = ord(ch) - ord("a")
            if not node.children[ch]:
                return None
            node = node.children[ch]
        return node

    def search(self, word: str) -> bool:
        node = self.search_prefix(word)
        return node is not None and node.is_end

    def starts_with(self, prefix: str) -> bool:
        return self.search_prefix(prefix) is not None


trie = Trie()
trie.insert("apple")
print(trie.search("apple"))  # 返回 True
print(trie.search("app"))  # 返回 False
print(trie.starts_with("app"))  # 返回 True
trie.insert("app")
print(trie.search("app"))  # 返回 True
